### DISCUSSION TITLE
TextLog, support of huge texts

### DISCUSSION THREAD

**Discussion by blademd** [Textualize/textual#2433] [Q&A]
Hello everyone!

I'm struggling with the TextLog when I need to draw a really huge text (i.e., several megabytes). The write() method freezes the app and after several minutes draws the text on a screen. After that, all work perfectly and I scroll through this text without any problems.

The only workaround I found is to put text piece by piece when the user is scrolling down. But I don't think it's an appropriate way to do that. (By the way, it would be nice to have an option to not auto scroll down a text with every write() call).

Is there something I'm missing? Or maybe there are some plans to enhance the TextLog?

Thanks a lot in advance. /Hope I found the right place to post this question/

**Comment by willmcgugan** [comment]
Freezing is more often due to blocking the event loop. If you post code, I might be able to give you suggestions.

See the [Textlog API Reference](https://textual.textualize.io/widgets/text_log/#textual.widgets.TextLog) for the `auto_scroll` attribute and `scroll_end`.

**Reply by blademd** [reply]
Hello!

I tried the simplest possible code:


```
from textual.app import App, ComposeResult
from textual.widgets import TextLog


class MyApp(App):
    def compose(self) -> ComposeResult:
        yield TextLog(highlight=True)

    def on_ready(self) -> None:
        text_log = self.query_one(TextLog)

        with open('conf', encoding='utf-8-sig') as f:
            data = f.read()
            text_log.write(data)


if __name__ == '__main__':
    app = MyApp()
    app.run()

```

The `conf` is 7636667 bytes long.

With the `highlight` option I got a much faster drawing time: 20-40 secs. in most cases. It's still too long but at least something interesting... Without this option, there are minutes of processing the output (based on the laptop fan's speed the stuff is really heavy).

This is the screenshot for the shortest case, the file' draw time is visible after the green line (I don't know why some timestamps after that line are the same as before but they appear in the log at the same time as the last ones):

![short](https://user-images.githubusercontent.com/1499024/235345660-a63e3b6a-ab87-4578-8367-8d907d3c2f0d.png)

And this one is for the longest case:

![long](https://user-images.githubusercontent.com/1499024/235345804-9512b737-51f2-444d-be6e-924e3ea02220.png)

20 secs. vs. 5+ mins.
